File: code/packages/typescript/chief-of-staff-channel-crypto/tests/d18f-fixtures.test.ts
import { readFileSync } from "node:fs";
import { generateKeypair } from "@coding-adventures/ed25519";
import { describe, expect, it } from "vitest";
import {
  D18Message,
  MAX_MESSAGE_JSON_BYTES,
  MessageProfileError,
  type MessageProfileErrorCode,
  MonotonicUuidV7Generator,
  messageAuthenticatedHeader,
  messageCreate,
  messageCreateWithSources,
  messageDeserialize,
  messageFromJson,
  messageSerialize,
  messageToJson,
  messageVerify,
  messageVerifyWithKeyResolver,
} from "../src/index.js";

interface PositiveCase {
  name: string;
  plaintext_b64: string;
  authenticated_header_b64: string;
  d18m_b64: string;
  canonical_json_b64: string;
}

interface NegativeBinaryCase {
  name: string;
  phase: "deserialize" | "verify";
  d18m_b64: string;
  expected_error: MessageProfileErrorCode;
}

interface NegativeJsonCase {
  name: string;
  json_b64: string;
  expected_error: MessageProfileErrorCode;
}

interface OversizeRecipe {
  field: "originator-id" | "content-type" | "ciphertext" | "json-input";
  declared_length: string;
  expected_error: MessageProfileErrorCode;
}

interface Fixture {
  fixture_format: string;
  generator_blob_sha1: string;
  warning: string;
  keys: {
    originator_signing_seed_hex: string;
    originator_public_key_hex: string;
    channel_master_keys: Array<{ key_epoch: string; key_hex: string }>;
  };
  positive_cases: PositiveCase[];
  binary_negative_cases: NegativeBinaryCase[];
  json_negative_cases: NegativeJsonCase[];
  oversize_recipes: OversizeRecipe[];
}

const fixture = JSON.parse(
  readFileSync(
    new URL(
      "../../../../fixtures/chief-of-staff-message/v1/manifest.json",
      import.meta.url,
    ),
    "utf8",
  ),
) as Fixture;

const signingSeed = fromHex(fixture.keys.originator_signing_seed_hex);
const { publicKey: derivedPublicKey, secretKey: signingSecretKey } =
  generateKeypair(signingSeed);
const publicKey = fromHex(fixture.keys.originator_public_key_hex);
const epochKeys = new Map(
  fixture.keys.channel_master_keys.map(({ key_epoch, key_hex }) => [
    BigInt(key_epoch),
    fromHex(key_hex),
  ]),
);

describe("D18F shared fixture", () => {
  it("locks fixture provenance and public test material", () => {
    expect(fixture.fixture_format).toBe("D18F-message-fixtures-v1");
    expect(fixture.generator_blob_sha1).toMatch(/^[0-9a-f]{40}$/);
    expect(fixture.warning).toContain("test-only");
    expect(fixture.positive_cases).toHaveLength(8);
    expect(fixture.binary_negative_cases).toHaveLength(20);
    expect(fixture.json_negative_cases).toHaveLength(11);
    expect(derivedPublicKey).toEqual(publicKey);
  });

  it("decodes, verifies, and reproduces every positive case byte-identically", () => {
    for (const testCase of fixture.positive_cases) {
      const binary = fromBase64(testCase.d18m_b64);
      const plaintext = fromBase64(testCase.plaintext_b64);
      const message = messageDeserialize(binary);
      const key = epochKeys.get(message.keyEpoch);
      expect(key, testCase.name).toBeDefined();

      expect(messageSerialize(message), testCase.name).toEqual(binary);
      expect(messageAuthenticatedHeader(message), testCase.name).toEqual(
        fromBase64(testCase.authenticated_header_b64),
      );
      expect(
        messageVerifyWithKeyResolver(message, publicKey, (epoch) => epochKeys.get(epoch)),
        testCase.name,
      ).toEqual(plaintext);
      expect(messageVerify(message, publicKey, key!), testCase.name).toEqual(plaintext);

      const canonicalJson = fromBase64(testCase.canonical_json_b64);
      expect(messageToJson(message), testCase.name).toEqual(canonicalJson);
      expect(messageSerialize(messageFromJson(canonicalJson)), testCase.name).toEqual(binary);

      const recreated = messageCreate(
        fieldsOf(message),
        plaintext,
        signingSecretKey,
        key!,
      );
      expect(messageSerialize(recreated), testCase.name).toEqual(binary);
    }
  });

  it("maps every binary mutation to its declared stable error", () => {
    for (const testCase of fixture.binary_negative_cases) {
      expectProfileError(() => {
        const message = messageDeserialize(fromBase64(testCase.d18m_b64));
        if (testCase.phase === "verify") {
          messageVerifyWithKeyResolver(message, publicKey, (epoch) => epochKeys.get(epoch));
        }
      }, testCase.expected_error, testCase.name);
    }
  });

  it("maps every JSON mutation to its declared stable error", () => {
    for (const testCase of fixture.json_negative_cases) {
      expectProfileError(
        () => messageFromJson(fromBase64(testCase.json_b64)),
        testCase.expected_error,
        testCase.name,
      );
    }
  });

  it("accepts arbitrary JSON field order and restores canonical order", () => {
    const canonical = new TextDecoder().decode(
      fromBase64(fixture.positive_cases[2].canonical_json_b64),
    );
    const reversed = JSON.stringify(
      Object.fromEntries(Object.entries(JSON.parse(canonical)).reverse()),
    );
    const restored = messageToJson(
      messageFromJson(new TextEncoder().encode(reversed)),
    );
    expect(new TextDecoder().decode(restored)).toBe(canonical);
  });

  it("rejects JSON strings that cannot represent lossless UTF-8", () => {
    const canonical = new TextDecoder().decode(
      fromBase64(fixture.positive_cases[0].canonical_json_b64),
    );
    const malformed = canonical.replace(
      '"content_type":"application/octet-stream"',
      '"content_type":"\\ud800"',
    );
    expectProfileError(
      () => messageFromJson(new TextEncoder().encode(malformed)),
      "invalid_field",
      "unpaired-surrogate",
    );
  });

  it("materializes compact oversize recipes without checked-in large blobs", () => {
    const baseline = fromBase64(fixture.positive_cases[0].d18m_b64);
    for (const recipe of fixture.oversize_recipes) {
      if (recipe.field === "json-input") {
        const logicalOversizeInput = { length: Number(recipe.declared_length) } as Uint8Array;
        expectProfileError(
          () => messageFromJson(logicalOversizeInput),
          recipe.expected_error,
          recipe.field,
        );
        continue;
      }

      const changed = baseline.slice();
      if (recipe.field === "originator-id") {
        setU32(changed, 29, Number(recipe.declared_length));
      } else if (recipe.field === "content-type") {
        setU32(changed, 83, Number(recipe.declared_length));
      } else {
        setU64(changed, 143, BigInt(recipe.declared_length));
      }
      expectProfileError(
        () => messageDeserialize(changed),
        recipe.expected_error,
        recipe.field,
      );
    }
    expect(Number(fixture.oversize_recipes[3].declared_length)).toBe(
      MAX_MESSAGE_JSON_BYTES + 1,
    );
  });
});

describe("TypeScript D18F API", () => {
  it("defensively copies constructor inputs and every byte accessor", () => {
    const source = messageDeserialize(fromBase64(fixture.positive_cases[1].d18m_b64));
    const parts = {
      ...fieldsOf(source),
      plaintextHash: source.plaintextHash,
      ciphertext: source.ciphertext,
      authenticationTag: source.authenticationTag,
      originatorSignature: source.originatorSignature,
    };
    const message = new D18Message(parts);
    const original = messageSerialize(message);

    parts.messageId.fill(0);
    parts.originatorId.fill(0);
    parts.channelId.fill(0);
    parts.plaintextHash.fill(0);
    parts.ciphertext.fill(0);
    parts.authenticationTag.fill(0);
    parts.originatorSignature.fill(0);
    message.messageId.fill(0);
    message.originatorId.fill(0);
    message.channelId.fill(0);
    message.plaintextHash.fill(0);
    message.ciphertext.fill(0);
    message.authenticationTag.fill(0);
    message.originatorSignature.fill(0);

    expect(Object.isFrozen(message)).toBe(true);
    expect(() => Object.assign(message, { sequence: 999n })).toThrow();
    expect(messageSerialize(message)).toEqual(original);
  });

  it("uses injected UUID-v7 and monotonic clock sources", () => {
    const source = messageDeserialize(fromBase64(fixture.positive_cases[0].d18m_b64));
    const key = epochKeys.get(source.keyEpoch)!;
    const message = messageCreateWithSources(
      {
        originatorId: source.originatorId,
        channelId: source.channelId,
        sequence: 123n,
        keyEpoch: source.keyEpoch,
        contentType: source.contentType,
      },
      new Uint8Array([1, 2, 3]),
      signingSecretKey,
      key,
      { next: () => source.messageId },
      { now: () => 456n },
    );
    expect(message.messageId).toEqual(source.messageId);
    expect(message.timestampNs).toBe(456n);
    expect(messageVerify(message, publicKey, key)).toEqual(new Uint8Array([1, 2, 3]));
  });

  it("generates 1,000 strictly ordered UUID-v7 values in one millisecond", () => {
    const generator = new MonotonicUuidV7Generator();
    let previous: Uint8Array | undefined;
    for (let index = 0; index < 1_000; index += 1) {
      const current = generator.next(1_725_000_000_000n, new Uint8Array(10).fill(0x55));
      expect(current[6] >> 4).toBe(7);
      expect(current[8] & 0xc0).toBe(0x80);
      if (previous !== undefined) expect(compareBytes(previous, current)).toBeLessThan(0);
      previous = current;
    }
  });
});

function fieldsOf(message: D18Message) {
  return {
    messageId: message.messageId,
    timestampNs: message.timestampNs,
    originatorId: message.originatorId,
    channelId: message.channelId,
    sequence: message.sequence,
    keyEpoch: message.keyEpoch,
    contentType: message.contentType,
  };
}

function expectProfileError(
  operation: () => unknown,
  code: MessageProfileErrorCode,
  label: string,
): void {
  try {
    operation();
    throw new Error(`${label}: expected ${code}`);
  } catch (error) {
    expect(error, label).toBeInstanceOf(MessageProfileError);
    expect((error as MessageProfileError).code, label).toBe(code);
  }
}

function fromHex(value: string): Uint8Array {
  return Uint8Array.from(value.match(/.{2}/g)?.map((byte) => Number.parseInt(byte, 16)) ?? []);
}

function fromBase64(value: string): Uint8Array {
  return new Uint8Array(Buffer.from(value, "base64"));
}

function setU32(bytes: Uint8Array, offset: number, value: number): void {
  new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength).setUint32(offset, value, false);
}

function setU64(bytes: Uint8Array, offset: number, value: bigint): void {
  new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength).setBigUint64(offset, value, false);
}

function compareBytes(left: Uint8Array, right: Uint8Array): number {
  for (let index = 0; index < left.length; index += 1) {
    if (left[index] !== right[index]) return left[index] - right[index];
  }
  return 0;
}

File: code/packages/elixir/chief-of-staff-channel-crypto/test/d18f_fixtures_test.exs
defmodule CodingAdventures.ChiefOfStaffChannelCrypto.D18FFixturesTest do
  use ExUnit.Case, async: true

  alias CodingAdventures.ChiefOfStaffChannelCrypto, as: Crypto

  alias Crypto.{
    D18Message,
    MessageFields,
    MessageProfileError,
    MonotonicUuidV7Generator,
    SourcedMessageFields
  }

  @fixture_path Path.expand(
                  "../../../../fixtures/chief-of-staff-message/v1/manifest.json",
                  __DIR__
                )
  @fixture @fixture_path |> File.read!() |> Jason.decode!()
  @signing_seed Base.decode16!(@fixture["keys"]["originator_signing_seed_hex"], case: :lower)
  @signing_keypair CodingAdventures.Ed25519.generate_keypair(@signing_seed)
  @public_key elem(@signing_keypair, 0)
  @signing_secret_key elem(@signing_keypair, 1)
  @expected_public_key Base.decode16!(@fixture["keys"]["originator_public_key_hex"], case: :lower)
  @epoch_keys Map.new(@fixture["keys"]["channel_master_keys"], fn item ->
                {String.to_integer(item["key_epoch"]),
                 Base.decode16!(item["key_hex"], case: :lower)}
              end)

  defp decode(value), do: Base.decode64!(value)

  defp assert_profile_error(code, operation) do
    error = assert_raise MessageProfileError, operation
    assert error.code == code
  end

  test "fixture provenance and public material are locked" do
    assert @fixture["fixture_format"] == "D18F-message-fixtures-v1"
    assert String.length(@fixture["generator_blob_sha1"]) == 40
    assert @fixture["warning"] =~ "test-only"
    assert length(@fixture["positive_cases"]) == 8
    assert length(@fixture["binary_negative_cases"]) == 20
    assert length(@fixture["json_negative_cases"]) == 11
    assert @public_key == @expected_public_key
  end

  test "positive fixtures are reproduced byte identically" do
    for test_case <- @fixture["positive_cases"] do
      binary = decode(test_case["d18m_b64"])
      plaintext = decode(test_case["plaintext_b64"])
      message = Crypto.message_deserialize(binary)
      key = Map.fetch!(@epoch_keys, message.key_epoch)

      assert Crypto.message_serialize(message) == binary, test_case["name"]

      assert Crypto.message_authenticated_header(message) ==
               decode(test_case["authenticated_header_b64"]),
             test_case["name"]

      assert Crypto.message_verify_with_key_resolver(
               message,
               @public_key,
               &Map.get(@epoch_keys, &1)
             ) == plaintext,
             test_case["name"]

      assert Crypto.message_verify(message, @public_key, key) == plaintext, test_case["name"]

      canonical = decode(test_case["canonical_json_b64"])
      assert Crypto.message_to_json(message) == canonical, test_case["name"]

      assert message
             |> Crypto.message_to_json()
             |> Crypto.message_from_json()
             |> Crypto.message_serialize() == binary

      recreated =
        Crypto.message_create(message_fields(message), plaintext, @signing_secret_key, key)

      assert Crypto.message_serialize(recreated) == binary, test_case["name"]
    end
  end

  test "binary mutations map to stable errors" do
    for test_case <- @fixture["binary_negative_cases"] do
      assert_profile_error(test_case["expected_error"], fn ->
        message = test_case["d18m_b64"] |> decode() |> Crypto.message_deserialize()

        if test_case["phase"] == "verify" do
          Crypto.message_verify_with_key_resolver(message, @public_key, &Map.get(@epoch_keys, &1))
        end
      end)
    end
  end

  test "JSON mutations map to stable errors" do
    for test_case <- @fixture["json_negative_cases"] do
      assert_profile_error(test_case["expected_error"], fn ->
        test_case["json_b64"] |> decode() |> Crypto.message_from_json()
      end)
    end
  end

  test "JSON field order is irrelevant and output is canonical" do
    canonical = decode(Enum.at(@fixture["positive_cases"], 2)["canonical_json_b64"])

    reversed =
      canonical
      |> Jason.decode!(objects: :ordered_objects)
      |> then(&%Jason.OrderedObject{values: Enum.reverse(&1.values)})
      |> Jason.encode!()

    assert reversed |> Crypto.message_from_json() |> Crypto.message_to_json() == canonical
  end

  test "JSON rejects unpaired surrogates" do
    canonical = decode(hd(@fixture["positive_cases"])["canonical_json_b64"])

    malformed =
      String.replace(
        canonical,
        ~s("content_type":"application/octet-stream"),
        ~s("content_type":"\\ud800")
      )

    assert_profile_error("invalid_field", fn -> Crypto.message_from_json(malformed) end)
  end

  test "JSON field types fail before magic semantics" do
    canonical = decode(hd(@fixture["positive_cases"])["canonical_json_b64"])
    malformed = String.replace(canonical, ~s("record_type":"D18M"), ~s("record_type":18))
    assert_profile_error("invalid_json", fn -> Crypto.message_from_json(malformed) end)
  end

  test "canonical JSON uses literal UTF-8 instead of optional escapes" do
    canonical = decode(hd(@fixture["positive_cases"])["canonical_json_b64"])

    escaped =
      String.replace(
        canonical,
        ~s("content_type":"application/octet-stream"),
        ~s("content_type":"application/\\u2028")
      )

    encoded = escaped |> Crypto.message_from_json() |> Crypto.message_to_json()
    assert encoded =~ "application/\u2028"
    refute encoded =~ "\\u2028"
  end

  test "compact oversize recipes are enforced" do
    baseline = decode(hd(@fixture["positive_cases"])["d18m_b64"])

    for recipe <- @fixture["oversize_recipes"] do
      if recipe["field"] == "json-input" do
        assert_profile_error(recipe["expected_error"], fn ->
          Crypto.message_from_json(:binary.copy(<<0>>, Crypto.max_message_json_bytes() + 1))
        end)
      else
        length = String.to_integer(recipe["declared_length"])

        changed =
          case recipe["field"] do
            "originator-id" -> replace_bytes(baseline, 29, 4, <<length::unsigned-big-32>>)
            "content-type" -> replace_bytes(baseline, 83, 4, <<length::unsigned-big-32>>)
            "ciphertext" -> replace_bytes(baseline, 143, 8, <<length::unsigned-big-64>>)
          end

        assert_profile_error(recipe["expected_error"], fn ->
          Crypto.message_deserialize(changed)
        end)
      end
    end
  end

  test "constructors validate boundaries and updates cannot mutate a message" do
    source =
      @fixture["positive_cases"]
      |> Enum.at(1)
      |> Map.fetch!("d18m_b64")
      |> decode()
      |> Crypto.message_deserialize()

    original = Crypto.message_serialize(source)
    changed = %{source | sequence: 999, ciphertext: <<0>>}
    assert Crypto.message_serialize(source) == original
    refute changed == source

    assert_profile_error("invalid_field", fn ->
      MessageFields.new!(
        <<0::120>>,
        source.timestamp_ns,
        source.originator_id,
        source.channel_id,
        source.sequence,
        source.key_epoch,
        source.content_type
      )
    end)

    assert_profile_error("invalid_field", fn ->
      D18Message.new!(source |> Map.from_struct() |> Map.put(:authentication_tag, <<0::120>>))
    end)
  end

  test "creation uses injected UUID and monotonic clock sources" do
    source =
      @fixture["positive_cases"]
      |> hd()
      |> Map.fetch!("d18m_b64")
      |> decode()
      |> Crypto.message_deserialize()

    key = Map.fetch!(@epoch_keys, source.key_epoch)

    fields =
      SourcedMessageFields.new!(
        source.originator_id,
        source.channel_id,
        123,
        source.key_epoch,
        source.content_type
      )

    message =
      Crypto.message_create_with_sources(
        fields,
        <<1, 2, 3>>,
        @signing_secret_key,
        key,
        fn -> source.message_id end,
        fn -> 456 end
      )

    assert message.message_id == source.message_id
    assert message.timestamp_ns == 456
    assert Crypto.message_verify(message, @public_key, key) == <<1, 2, 3>>
  end

  test "UUID-v7 generator orders 1000 values in one millisecond" do
    {_generator, _previous} =
      Enum.reduce(1..1000, {MonotonicUuidV7Generator.new(), nil}, fn _, {generator, previous} ->
        {generator, current} =
          MonotonicUuidV7Generator.next(generator, 1_725_000_000_000, :binary.copy(<<0x55>>, 10))

        <<_::binary-size(6), version::4, _::12, variant::2, _::6, _::binary>> = current
        assert version == 7
        assert variant == 2
        if previous != nil, do: assert(previous < current)
        {generator, current}
      end)
  end

  defp message_fields(message) do
    MessageFields.new!(
      message.message_id,
      message.timestamp_ns,
      message.originator_id,
      message.channel_id,
      message.sequence,
      message.key_epoch,
      message.content_type
    )
  end

  defp replace_bytes(value, offset, length, replacement) do
    <<prefix::binary-size(offset), _::binary-size(length), suffix::binary>> = value
    prefix <> replacement <> suffix
  end
end
